nvflare/app_common/dex/dxo_exchanger.py

```diff
@@ -0,0 +1,115 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import time
+from typing import Optional
+
+from nvflare.apis.dxo import DXO
+from nvflare.fuel.utils.pipe.file_accessor import FileAccessor
+from nvflare.fuel.utils.pipe.file_pipe import FilePipe
+from nvflare.fuel.utils.pipe.pickle_file_accessor import PickleFileAccessor
+from nvflare.fuel.utils.pipe.pipe import Message
+from nvflare.fuel.utils.pipe.pipe_monitor import PipeMonitor, Topic
+
+
+class DXOExchanger:
+    def __init__(
+        self,
+        pipe_role: str,
+        pipe_name: str = "pipe",
+        get_poll_interval: float = 0.1,
+        heartbeat_interval: float = 0.5,
+        heartbeat_timeout: float = 0.0,
+    ):
+        """Exchanges DXO.
+
+        Args:
+            pipe_role (str): Should be either "x", or "y". A pipe has two endpoints, one side should be "x" while
+                the other side should be "y".
+            pipe_name (str): Name of the pipe, default is "pipe".
+            get_poll_interval (float): how often to check if the other side has sent file
+            heartbeat_interval (float): how often to send a heartbeat to the peer
+            heartbeat_timeout (float): how long to wait for a heartbeat from the peer before treating the peer as gone,
+                0 means DO NOT check for heartbeat.
+        """
+        self.logger = logging.getLogger(self.__class__.__name__)
+        self._get_poll_interval = get_poll_interval
+
+        self.data_exchange_path = None
+        self._pipe_role = pipe_role
+        self._pipe_name = pipe_name
+        self._heartbeat_interval = heartbeat_interval
+        self._heartbeat_timeout = heartbeat_timeout
+        self.pipe_monitor = None
+        self._file_accessor_name = None
+
+    def initialize(self, data_exchange_path: str, file_accessor: Optional[FileAccessor] = None):
+        self.data_exchange_path = os.path.abspath(data_exchange_path)
+        file_pipe = FilePipe(self.data_exchange_path)
+        if file_accessor is not None:
+            file_pipe.set_file_accessor(file_accessor)
+        else:
+            file_pipe.set_file_accessor(PickleFileAccessor())
+
+        # use file accessor name to determine serialize/deserialize
+        self._file_accessor_name = file_pipe.accessor.__class__.__name__
+        file_pipe.open(self._pipe_name, self._pipe_role)
+        self.pipe_monitor = PipeMonitor(
+            file_pipe,
+            read_interval=self._get_poll_interval,
+            heartbeat_interval=self._heartbeat_interval,
+            heartbeat_timeout=self._heartbeat_timeout,
+        )
+        self.pipe_monitor.start()
+
+    def put(self, data_id: str, data: DXO):
+        if self.pipe_monitor is None:
+            raise RuntimeError("PipeMonitor is not initialized.")
+        req = Message.new_request(self._file_accessor_name, data, data_id)
```

Comment:
The topic should not be self._file_accessor_name, which should not be exposed at all.  
We should get rid of the term "data_id" and replace it with "topic".

Reply:
thanks, changed.